### PR TITLE
Consensus project tests part 1

### PR DESCRIPTION
**Pull request by Neurosploit**
I've started to create tests for the consensus project. 
This commit contains tests for some general classes of the coinviews folder, consensus rule executers and a start to unit tests for the consensus rules. The plan is to add tests like dan added for consensusrules later after the unit tests are in place.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Consensus.Tests/Rules/ConsensusRuleUnitTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            this.loggerFactory.Setup(l => l.CreateLogger(It.IsAny<string>()))
+                .Returns(new Mock<ILogger>().Object);
+            this.dateTimeProvider = new Mock<IDateTimeProvider>();
+
+            this.concurrentChain = new ConcurrentChain(this.network);
+            this.nodeDeployments = new NodeDeployments(this.network, this.concurrentChain);
+            this.consensusSettings = new ConsensusSettings
+            {
+                BlockAssumedValid = null,
+                UseCheckpoints = true
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus.Tests/Rules/ConsensusRuleUnitTestBase.cs]: I wonder if this should be true by default

File: src/Stratis.Bitcoin.Features.Consensus.Tests/Stratis.Bitcoin.Features.Consensus.Tests.csproj [hunk 2]
```diff
@@ -35,6 +35,7 @@
     <ProjectReference Include="..\NBitcoin\NBitcoin.csproj" />
     <ProjectReference Include="..\Stratis.Bitcoin.Features.Consensus\Stratis.Bitcoin.Features.Consensus.csproj" />
     <ProjectReference Include="..\Stratis.Bitcoin.Features.Miner\Stratis.Bitcoin.Features.Miner.csproj" />
+    <ProjectReference Include="..\Stratis.Bitcoin.Features.Wallet.Tests\Stratis.Bitcoin.Features.Wallet.Tests.csproj" />
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus.Tests/Stratis.Bitcoin.Features.Consensus.Tests.csproj]: I think this is wrong wallet depends on consensus not the other way around.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs [hunk 3]
```diff
@@ -39,10 +42,14 @@ public IEnumerable<CoinView> GetElements()
             CoinView current = this.Top;
             while (current is IBackedCoinView)
             {
-                yield return current;
+                if (current != null)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs]: This failed in one of the tests?

File: src/Stratis.Bitcoin.Features.MemoryPool/MemPoolCoinView.cs [hunk 4]
```diff
@@ -86,7 +86,7 @@ public override Task<uint256> Rewind()
                 return this.memPool.MapTx.Values.Where(t => ids.Contains(t.TransactionHash)).Select(s => s.Transaction).ToList();
             });
             IEnumerable<UnspentOutputs> memOutputs = mempoolcoins.Select(s => new UnspentOutputs(TxMempool.MempoolHeight, s));
-            coins.UnspentOutputs = coins.UnspentOutputs.Concat(memOutputs).ToArray();
+            coins = new FetchCoinsResponse(coins.UnspentOutputs.Concat(memOutputs).ToArray(), coins.BlockHash);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.MemoryPool/MemPoolCoinView.cs]: You think this is better? well why not.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs (hunk 3 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs]: I got a null value at the end of the list.

File: src/Stratis.Bitcoin.Features.MemoryPool/MemPoolCoinView.cs (hunk 4 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.MemoryPool/MemPoolCoinView.cs]: it prevents null assignment to that member

File: src/Stratis.Bitcoin.Features.Consensus.Tests/Stratis.Bitcoin.Features.Consensus.Tests.csproj (hunk 2 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus.Tests/Stratis.Bitcoin.Features.Consensus.Tests.csproj]: Ok I will fix this.

File: src/Stratis.Bitcoin.Features.Consensus.Tests/Rules/ConsensusRuleUnitTestBase.cs (hunk 1 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus.Tests/Rules/ConsensusRuleUnitTestBase.cs]: Fair enough. I will move that out. I will move that out and fix the test that needs it.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs (hunk 3 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinViewStack.cs]: Edit: I got a null value in the ienumerable from the end of the coinviews stack not having an inner. I will check if we truly need the null check inside the while loop. I think that one can be removed but keep the one before the end.

**Comment by timhoekstra** [comment]
@dangershony please have another look when you have time. Thanks.

**Comment by Neurosploit** [comment]
@dangershony is this ok now or do I need to change anything else?

**Comment by dangershony** [comment]
Good job this can go in now.

**Pull request merged**